### ISSUE TITLE
Setting instance field of Callable type with the return value of calling a generic function causes errors

### ISSUE THREAD

**Issue by rshin** [microsoft/pyright#2965]
**Describe the bug**
See code below.

**To Reproduce**
See code below.

**Expected behavior**
`c.str_to_int = decorate(f)` in the code below does not cause any problems.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from dataclasses import dataclass
from typing import Callable, TypeVar

CallableT = TypeVar("CallableT", bound=Callable)

def decorate(arg: CallableT) -> CallableT:
    return arg

def f(s: str) -> int:
    return int(s)

# Type of "decorate(f)" is "(s: str) -> int"
reveal_type(decorate(f))

@dataclass
class C:
    str_to_int: Callable[[str], int] = f

c = C()

## Works fine
c.str_to_int = f

## Causes errors
c.str_to_int = decorate(f)

## Works fine
f_d = decorate(f)
# Type of "f_d" is "(s: str) -> int"
reveal_type(f_d)
c.str_to_int = f_d
```

**VS Code extension or command-line**
Pylance v2022.1.5 in VS Code.
Same behavior in pyright 1.1.216.

**Comment by erictraut** [comment]
Thanks for the bug report. Fields within a dataclass require special casing because they are instance variables rather than class variables, so they don't get bound to the object when accessed through a member access expression (e.g. `c.str_to_int`). This special-case logic was missing in one code path. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.217, which I just published. It will also be included in the next release of pylance.
